### ISSUE TITLE
Command callback triggers with invalid command context

### ISSUE THREAD

**Issue by chris-perry** [FreeRTOS/Lab-Project-coreMQTT-Agent#33]
# Issue Description
I am experiencing an issue where a command callback will attempt to access data from a command that is out of context, and in our application, this triggers a hard fault. 

I've read over the agent documentation listed at https://freertos.org/mqtt/mqtt-agent-demo.html#example_mqtt_api_call
and our application essentially uses the agent implementation presented in the simple sub pub demo. So I'm not entirely sure how the command would be getting out of context since those tasks are waiting on a task notify to continue. 

The application uses the kernel and other libraries from the FreeRTOS 202012 LTS release in addition with the cellular interface and coremqtt agent libraries.

# Reproducing with windows demo
I believe we can induce similar behavior with the coreMQTT Agent windows simple pub sub demo (  64df467 )

Notable changes, See diff at the bottom:
* mqttexampleMS_TO_WAIT_FOR_NOTIFICATION, to 100 ms
* democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE to 6 
* Only trigger configASSERT(ulNotification == ulValueToNotify) if prvWaitForCommandAcknowledgment is successful.

```diff
diff --git a/source/configuration-files/demo_config.h b/source/configuration-files/demo_config.h
index e906e30..3cbbb53 100644
--- a/source/configuration-files/demo_config.h
+++ b/source/configuration-files/demo_config.h
@@ -74,17 +74,17 @@ extern void vLoggingPrintf( const char * pcFormatString,
 #define democonfigCREATE_LARGE_MESSAGE_SUB_PUB_TASK        0
 #define democonfigLARGE_MESSAGE_SUB_PUB_TASK_STACK_SIZE    ( configMINIMAL_STACK_SIZE )
 
-#define democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE       0
+#define democonfigNUM_SIMPLE_SUB_PUB_TASKS_TO_CREATE       4
 #define democonfigSIMPLE_SUB_PUB_TASK_STACK_SIZE           ( configMINIMAL_STACK_SIZE )
 
 #define democonfigCREATE_CODE_SIGNING_OTA_DEMO             1
 #define democonfigCODE_SIGNING_OTA_TASK_STACK_SIZE         ( configMINIMAL_STACK_SIZE )
 
 
-#define democonfigCREATE_DEFENDER_DEMO                     1
+#define democonfigCREATE_DEFENDER_DEMO                     0
 #define democonfigDEFENDER_TASK_STACK_SIZE                 ( configMINIMAL_STACK_SIZE )
 
-#define democonfigCREATE_SHADOW_DEMO                       1
+#define democonfigCREATE_SHADOW_DEMO                       0
 #define democonfigSHADOW_TASK_STACK_SIZE                   ( configMINIMAL_STACK_SIZE )
 
 /**
diff --git a/source/simple_sub_pub_demo.c b/source/simple_sub_pub_demo.c
index 3d8b3ad..bc152b7 100644
--- a/source/simple_sub_pub_demo.c
+++ b/source/simple_sub_pub_demo.c
@@ -68,7 +68,7 @@
  * functions.  mqttexampleMS_TO_WAIT_FOR_NOTIFICATION defines the time, in ticks,
  * to wait for such a callback.
  */
-#define mqttexampleMS_TO_WAIT_FOR_NOTIFICATION            ( 5000 )
+#define mqttexampleMS_TO_WAIT_FOR_NOTIFICATION            ( 100 )
 
 /**
  * @brief Size of statically allocated buffers for holding topic names and
@@ -437,10 +437,11 @@ static void prvSimpleSubscribePublishTask( void * pvParameters )
     TickType_t xTicksToDelay;
     CommandInfo_t xCommandParams = { 0 };
     char * pcTopicBuffer = topicBuf[ ulTaskNumber ];
+    BaseType_t xValidNotification = false;
 
     /* Have different tasks use different QoS.  0 and 1.  2 can also be used
      * if supported by the broker. */
-    xQoS = ( MQTTQoS_t ) ( ulTaskNumber % 2UL );
+    xQoS = ( MQTTQoS_t ) ( 1 );
 
     /* Create a unique name for this task from the task number that is passed into
      * the task using the task's parameter. */
@@ -507,12 +508,16 @@ static void prvSimpleSubscribePublishTask( void * pvParameters )
         LogInfo( ( "Task %s waiting for publish %d to complete.",
                    taskName,
                    ulValueToNotify ) );
-        prvWaitForCommandAcknowledgment( &ulNotification );
+
+        xValidNotification = prvWaitForCommandAcknowledgment( &ulNotification );
 
         /* The value by the callback that executed when the publish was acked
          * came from the context passed into MQTTAgent_Publish() above, so
          * should match the value set in the context above. */
-        configASSERT( ulNotification == ulValueToNotify );
+        if (xValidNotification)
+        {
+            configASSERT(ulNotification == ulValueToNotify);
+        }
 
         if( ulNotification == ulValueToNotify )
         {
```

With these modifications, one of the following issues will present itself within a couple of minutes:

Issue 1 - Spurious invalid notify task on prvSubscribeCommandCallback: 
If you set mqttexampleMS_TO_WAIT_FOR_NOTIFICATION to 10 ms then this gets triggered immediately

Issue 2 -  configASSERT(ulNotification == ulValueToNotify) triggered in subpub task:
This issue will always occur after a command times out.

Sequence is as follows
*  Subpub task waits for command to execute
*  prvWaitForCommandAcknowledgment times out
*  that specific subpub task waits for the next command 
*  prvWaitForCommandAcknowledgment will return with the wrong notify value 

I was expecting that after an MQTT command timeout occurs any references to that command would have been 
invalidated/ignored by the agent.

**Comment by muneebahmed10** [comment]
Hi @chris-perry,

At the moment, there isn't really a concept of command timeout with the MQTT agent, since no timeouts are passed when creating the command. The block time that is passed is only for adding the command to the queue. The timeout values from `mqttexampleMS_TO_WAIT_FOR_NOTIFICATION` are just example durations that would be reasonable to expect for any command to complete. Subscribes and publishes with QoS > 0 expect acknowledgment packets, so the command is only considered complete once it receives the ack; this means you need to wait for more than the time for a network round trip, and 10ms may not be enough.

Since commands do not have timeouts, the only way to know when it has completed is when the completion callback has been invoked (if one was passed in). If the completion callback is not invoked in a reasonable amount of time, you can perform the following steps:
1. Call MQTTAgent_Terminate() from any task. This will add a command to terminate the loop, and will also invoke completion callbacks for all commands currently in the queue - and those awaiting acknowledgments - with an error code.
2. Restart the command loop from the agent task with [MQTTAgent_CommandLoop()](https://github.com/FreeRTOS/Lab-Project-coreMQTT-Agent/blob/main/source/connection_manager.c#L863). The code for the agent task in `connection_manager.c` is just an example demo of how you may write an agent task. You can, of course, modify it so that the loop is restarted after MQTTAgent_Terminate() is called.

Additionally, if you don't need to use QoS 1, then QoS 0 publishes will complete much faster, as the command completes as soon as the publish is sent.

We may consider adding command timeouts to the MQTT Agent in a future release. I will make note of your feedback and bring this up in our planning.

Let me know if you have any further questions.

**Comment by muneebahmed10** [comment]
Hi @chris-perry,

I will close this issue. Please open a new issue if you have any other questions.

Thank you
